### PR TITLE
datapath: Enable N/S LB for overlapping pod CIDR

### PR DESCRIPTION
**Pull request by jibi**
see individual commits

### CODE REVIEW

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -2709,6 +2716,10 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 
 		src_port = tunnel_gen_src_port_v4(&tuple);
 
+#if !(defined(ENABLE_CLUSTER_AWARE_ADDRESSING) && defined(ENABLE_INTER_CLUSTER_SNAT))
```
Comment (jibi) [inline bpf/lib/nodeport.h]: offline it was discussed to have the callers of `nodeport_lb4()` set the `src_sec_identity`, and then avoid messing with it inside the tail calls chain (@brb) but this needs to remain in `tail_nodeport_nat_egress_ipv4()` as overwriting the source identity with `WORLD_IPV4_ID` in the callers would break policy enforcement

File: bpf/bpf_host.c [hunk 2]
```diff
@@ -880,6 +880,9 @@ tail_handle_ipv4(struct __ctx_buff *ctx, __u32 ipcache_srcid, const bool from_ho
 	int ret;
 	__s8 ext_err = 0;
 
+	if (src_sec_identity == 0)
+		src_sec_identity = WORLD_IPV4_ID;
+
```
Comment (julianwiedmann) [inline bpf/bpf_host.c]: Imho this change (and the corresponding one in bpf_overlay) is way too broad, and can have unexpected side-effects for existing code. Why not push this "override" logic down into `nodeport_lb4()` ?

File: bpf/lib/nodeport.h [hunk 3]
```diff
@@ -2928,7 +2928,7 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 		switch (ret) {
 		case CT_NEW:
 redo:
-			ct_state_new.src_sec_id = WORLD_IPV4_ID;
+			ct_state_new.src_sec_id = src_sec_identity;
 			ct_state_new.node_port = 1;
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h:2944]: Minor patch ordering aspect - let's move this patch a bit further back in the series, after all the "set src identity to world if 0" changes are done?

I'm not aware that any BPF code uses the `src_sec_id` (it's definitely not returned on the CT lookup), did you check how this change would affect the agent/proxy?

File: bpf/lib/nodeport.h [hunk 4]
```diff
@@ -2661,6 +2662,8 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
 	struct iphdr *ip4;
 	__s8 ext_err = 0;
 #ifdef TUNNEL_MODE
+	__u8 cluster_id __maybe_unused = (__u8)ctx_load_meta(ctx, CB_CLUSTER_ID_EGRESS);
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h]: The `__maybe_unused` isn't needed, correct?

File: bpf/lib/nodeport.h [hunk 5]
```diff
@@ -2661,6 +2662,8 @@ int tail_nodeport_nat_egress_ipv4(struct __ctx_buff *ctx)
...
+	__u32 src_identity = ctx_load_meta(ctx, CB_SRC_LABEL);
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h]:
```suggestion
	__u32 src_sec_identity = ctx_load_meta(ctx, CB_SRC_LABEL);
```

File: bpf/lib/nodeport.h [hunk 6]
```diff
@@ -2979,6 +2986,8 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
 #endif /* DSR_ENCAP_MODE */
 		ep_tail_call(ctx, CILIUM_CALL_IPV4_NODEPORT_DSR);
 	} else {
+		ctx_store_meta(ctx, CB_SRC_LABEL, src_sec_identity);
+		ctx_store_meta(ctx, CB_CLUSTER_ID_EGRESS, cluster_id);
```
Comment (julianwiedmann) [inline bpf/lib/nodeport.h:2995]: FWIW, I think it would be worth doing separate patches for the `cluster_id` and `src_sec_identity` change. They are handling different aspects - and while the `cluster_id` change is trivial, the `src_sec_identity` thing has more potential to need a revert.

File: bpf/lib/encap.h [hunk 7]
```diff
@@ -32,6 +34,9 @@ __encap_with_nodeid(struct __ctx_buff *ctx, __u32 src_ip, __be16 src_port,
 	if (seclabel == HOST_ID)
 		seclabel = LOCAL_NODE_ID;
 
+	if (seclabel > ENCAP_MAX_VNI)
+		return DROP_INVALID_IDENTITY;
+
```
Comment (julianwiedmann) [inline bpf/lib/encap.h]: To clarify - is this patch meant to address the CIDR-identity concern from https://github.com/cilium/cilium/pull/29923#discussion_r1429574175, or are you planning to add more graceful handling?

**Review by julianwiedmann — CHANGES_REQUESTED** [review]
Thank you! A few change requests below.

File: bpf/bpf_host.c (hunk 2 above)
Comment (jibi) [reply bpf/bpf_host.c]: see my previous comment, @brb asked offline to have the caller of `nodeport_lb4()` set the src identity so I just followed that, but I don't mind moving it a bit deeper in the call stack

File: bpf/lib/nodeport.h (hunk 3 above)
Comment (jibi) [reply bpf/lib/nodeport.h:2944]: >I'm not aware that any BPF code uses the src_sec_id (it's definitely not returned on the CT lookup), did you check how this change would affect the agent/proxy?

yes, it just used for debugging purposes and printed when the CT table is dumped

File: bpf/lib/nodeport.h (hunk 4 above)
Comment (jibi) [reply bpf/lib/nodeport.h]: it's needed in case `defined(ENABLE_CLUSTER_AWARE_ADDRESSING) && defined(ENABLE_INTER_CLUSTER_SNAT)` is not true as that guards the only block where cluster_id is used

File: bpf/lib/encap.h (hunk 7 above)
Comment (jibi) [reply bpf/lib/encap.h]: for now just to make sure we don't pass down local identities. This is the only place where we return the invalid ID drop reason, so should be already a good way to detect if this happens

File: bpf/lib/nodeport.h (hunk 4 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h]: Ah, otherwise we still pass it to `lookup_ip4_remote_endpoint()` but then no-one uses the result :+1: .

File: bpf/bpf_host.c (hunk 2 above)
Comment (julianwiedmann) [reply bpf/bpf_host.c]: hm yes, but you're changing the calling paths at a much higher level - it's not obvious that **all** subsequent users of the src_sec_identity want to have this override. What if one of them has unique handling for `identity == 0`, that now doesn't get applied any more?

I think you'll have a much easier life keeping this contained inside the nodeport-scope for now.

Comment (jibi) [reply bpf/bpf_host.c]: > hm yes, but you're changing the calling paths at a much higher level

yes, on purpose in order to have a consistent value across all users rather than having it set to 0 in one code path and WORLD_IPV4_ID in another

> it's not obvious that all subsequent users of the src_sec_identity want to have this override. What if one of them has unique handling for identity == 0, that now doesn't get applied any more?

fair, perhaps I have too much faith in the tests :disappointed: 

> I think you'll have a much easier life keeping this contained inside the nodeport-scope for now.

sounds good, I'll go with that then :ok_hand:

File: bpf/lib/nodeport.h (hunk 6 above)
Comment (julianwiedmann) [reply bpf/lib/nodeport.h:2995]: Sorry, I should have been clearer. I meant to extract *all* of the `cluster_id` related changes into a separate patch. Not just this single line.

File: bpf/bpf_host.c (hunk 2 above)
Comment (julianwiedmann) [reply bpf/bpf_host.c]: Just checking - this isn't updated in the latest push, right? The identity override is still in every calling path, not down in `nodeport_lb4()`.

File: bpf/lib/encap.h (hunk 7 above)
Comment (julianwiedmann) [reply bpf/lib/encap.h]: But you could easily catch that at the specific location where `Cluster-aware-addressing` sets the forwarded identity? And thus self-document that it's not supported yet (and why).

File: bpf/bpf_host.c (hunk 2 above)
Comment (jibi) [reply bpf/bpf_host.c]: ah, I understood we just wanted to move it right before calling into `nodeport_lb4()` and not in that function, let me move it back down to that function then

**Review by brb — APPROVED** [review]
Thanks!

**Review by julianwiedmann — APPROVED** [review]
diff looks good, ty jibi!

File: bpf/lib/nodeport.h [hunk 8]
```diff
@@ -2907,6 +2914,22 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
...
+#endif
+
+		 /* Before forwarding the identity, make sure it's not a CIDR
+		  * identity, as these are __u32 values, but transporting them
+		  * via the VNI field in the VXLAN / Geneve header allows for
+		  * only 24 bits.
+		  */
+
+		if (identity_is_cidr_range(src_sec_identity))
+			return DROP_INVALID_IDENTITY;
```
Comment (joestringer) [inline bpf/lib/nodeport.h:2931]: Does this make sense?

If I follow correctly, a packet coming into a NodePort from outside of the cluster could be classified by Cilium to have a CIDR identity as the source, which is completely fine, then instead of forwarding the request, we just drop it?

File: bpf/lib/nodeport.h [hunk 9]
```diff
@@ -2907,6 +2914,22 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
...
+#else
+		if (src_sec_identity == 0)
+			src_sec_identity = WORLD_IPV4_ID;
+#endif
+
+		 /* Before forwarding the identity, make sure it's not a CIDR
+		  * identity, as these are __u32 values, but transporting them
+		  * via the VNI field in the VXLAN / Geneve header allows for
+		  * only 24 bits.
+		  */
```
Comment (joestringer) [inline bpf/lib/nodeport.h:2928]: I'm not sure that the size of the identity is the problem here. The problem is that the identity we are trying to encode into the vxlan header only has meaning locally. So even if it were possible to somehow encode the identity into the header, once the packet arrives at the next hop, that hop wouldn't understand the identity because it's not synchronized cluster-wide (or across clusters).

File: bpf/lib/nodeport.h [hunk 10]
```diff
@@ -2907,6 +2914,22 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
...
+
+		if (identity_is_cidr_range(src_sec_identity))
```
Comment (joestringer) [inline bpf/lib/nodeport.h:2930]: I wonder also whether we shouldn't just be checking "in CIDR range", but actually the full upper 8 bits check (are any bits there set). Comparing this with `pkg/identity/numericidentity.go` "scopes" I also see the remote node scope, which could use a range from `2<<24` up , which would be outside the range that this function checks :thinking:

File: bpf/lib/nodeport.h (hunk 8 above)
Comment (joestringer) [reply bpf/lib/nodeport.h:2931]: Hmm, I guess if we think this should never happen because of other constraints in the datapath (ie we never expect such identities to show up at this point in the processing), then as a way to do a safety check this may make sense. We may want to add `HOST_ID` to these as well, since we really shouldn't forward that identity on to another node (nor should another node blindly trust an incoming usage of the host identity).

File: bpf/lib/nodeport.h (hunk 9 above)
Comment (jibi) [reply bpf/lib/nodeport.h:2928]: yeah, I think I'm just quoting the implementation detail (CIDR identities aka local identities don't fit in the VNI field) rather than stating clearly that forwarding local identities doesn't make sense, but that's the base idea
